src/flamegraph/color.rs

```diff
@@ -0,0 +1,272 @@
+                if suffix_char.starts_with("j") { return Palette::Green }
+            }
+        }
+    }
+
+    let java_prefix = if s.starts_with("L") { &s[1..] } else { s };
+
+    if java_prefix.starts_with("java/") ||
+        java_prefix.starts_with("org/") ||
+        java_prefix.starts_with("com/") ||
+        java_prefix.starts_with("io/") ||
+        java_prefix.starts_with("sun/") {
+        Palette::Green
+    } else if s.contains("::") {
+        Palette::Yellow
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_perl_palette(s: &str) -> Palette {
+    if s.contains("::") {
+        Palette::Yellow
+    } else if s.contains("Perl") || s.contains(".pl") {
+        Palette::Green
+    } else if s.ends_with("_[k]") {
+        Palette::Orange
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_js_palette(s: &str) -> Palette {
+    if s.ends_with("_[j]") {
+        if s.contains("/") {
+            return Palette::Green
+        } else {
+            return Palette::Aqua
+        }
+    } else if s.contains("::") {
+        return Palette::Yellow
+    } else if let Some(ai) = s.find("/") {
+        if (&s[ai..]).contains(".js") {
+            return Palette::Green
+        }
+    }
+
+    if s.contains(":") {
+        Palette::Aqua
+    } else if s == " " {
+        Palette::Green
+    } else if s.contains("_[k]") {
+        Palette::Orange
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_wakeup_palette(_s: &str) -> Palette {
+    Palette::Aqua
+}
+
+fn handle_chain_palette(s: &str) -> Palette {
+    if s.contains("_[w]") {
+        Palette::Aqua
+    } else {
+        Palette::Blue
+    }
+}
+
+fn coefficients_for_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) ->
+    ((u8, u8, f32), (u8, u8, f32), (u8, u8, f32)) {
+    let real_palette = match palette {
+        Palette::Hot => return ((205, 50, v3), (0, 230, v1), (0, 55, v2)),
+        Palette::Mem => return ((0, 0, v3), (190, 50, v2), (0, 210, v1)),
+        Palette::Io => return ((80, 60, v1), (80, 60, v1), (190, 55, v2)),
+        Palette::Red => return ((200, 55, v1), (50, 80, v1), (50, 80, v1)),
+        Palette::Green => return ((50, 60, v1), (200, 55, v1), (50, 60, v1)),
+        Palette::Blue => return ((80, 60, v1), (80, 60, v1), (205, 60, v1)),
+        Palette::Yellow => return ((175, 55, v1), (175, 55, v1), (50, 20, v1)),
+        Palette::Purple => return ((190, 65, v1), (80, 60, v1), (190, 65, v1)),
+        Palette::Aqua => return ((50, 60, v1), (165, 55, v1), (165, 55, v1)),
+        Palette::Orange => return ((190, 65, v1), (90, 65, v1), (0, 0, v1)),
+        Palette::Java => handle_java_palette(name),
+        Palette::Perl => handle_perl_palette(name),
+        Palette::Js => handle_js_palette(name),
+        Palette::Wakeup => handle_wakeup_palette(name),
+        Palette::Chain => handle_chain_palette(name),
+    };
+
+    coefficients_for_palette(&real_palette, name, v1, v2, v3)
```

Comment:
This confused me for a second -- I think a comment describing what this is for (resolving palettes used by language palettes) would be helpful.

Reply:
It may be clearer if we separate language palettes from "basic" palettes in the ` Palette` enum, with a function resolving only the language palette enum into the basic palette enum ?

Reply:
Ah, yes, I like that idea!

Reply:
[Done](https://github.com/jonhoo/inferno/pull/32/commits/e25e449aa9208bf5b3715b5bfec33b3c5b01a031)